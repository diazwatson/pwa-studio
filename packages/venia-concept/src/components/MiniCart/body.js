import React, { useCallback } from 'react';
import { array, bool, func, object, shape, string } from 'prop-types';

import { mergeClasses } from 'src/classify';
import { loadingIndicator } from 'src/components/LoadingIndicator';

import defaultClasses from './body.css';
import EditItem from './editItem';
import EmptyMiniCartBody from './emptyMiniCartBody';
import ProductList from './productList';

const Body = props => {
    // Props.
    const {
        beginEditItem,
        cart,
        closeDrawer,
        endEditItem,
        isCartEmpty,
        isMiniCartMaskOpen,
        removeItemFromCart,
        updateItemInCart
    } = props;

    // Members.
    const classes = mergeClasses(defaultClasses, props.classes);
    const { editItem, isLoading, isEditingItem, isUpdatingItem } = cart;

    // Callbacks.
    const handleEditItem = useCallback(
        item => {
            beginEditItem(item);
        },
        [beginEditItem]
    );

    // Render.
    if (isLoading) {
        return loadingIndicator;
    }

    if (isCartEmpty) {
        return <EmptyMiniCartBody closeDrawer={closeDrawer} />;
    }

    if (isEditingItem) {
        return (
            <EditItem
                endEditItem={endEditItem}
                isUpdatingItem={isUpdatingItem}
                item={editItem}
                updateItemInCart={updateItemInCart}
            />
        );
    }

    return (
        <div className={classes.root}>
            <ProductList
                beginEditItem={handleEditItem}
                cart={cart}
                isMiniCartMaskOpen={isMiniCartMaskOpen}
                items={cart.details.items}
                removeItemFromCart={removeItemFromCart}
                totalsItems={cart.totals.items}
            />
        </div>
    );
};

Body.propTypes = {
    beginEditItem: func.isRequired,
    cart: shape({
        details: shape({
            items: array
        }).isRequired,
        editItem: object,
        isEditingItem: bool,
        isLoading: bool,
        isUpdatingItem: bool,
        totals: shape({
            items: array
        }).isRequired
    }).isRequired,
    classes: shape({
        root: string
    }),
    closeDrawer: func,
    endEditItem: func,
    isCartEmpty: bool,
    isMiniCartMaskOpen: bool,
    removeItemFromCart: func,
    updateItemInCart: func
};

export default Body;
