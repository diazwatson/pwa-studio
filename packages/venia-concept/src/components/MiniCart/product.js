import React, { useCallback, useEffect, useRef, useState } from 'react';
import { array, func, number, shape, string } from 'prop-types';
import { Price } from '@magento/peregrine';

import { mergeClasses } from 'src/classify';
import { resourceUrl } from 'src/drivers';

import Kebab from './kebab';
import ProductOptions from './productOptions';
import Section from './section';

import defaultClasses from './product.css';

const FAVORITES_FILL = { fill: 'rgb(var(--venia-teal))' };

const Product = props => {
    const { beginEditItem, currencyCode, item, removeItemFromCart } = props;
    const { image, name, options, price, qty } = item;

    const imageRef = useRef(null);
    const [imageSource, setImageSource] = useState(null);
    const [isFavorite, setIsFavorite] = useState(false);
    const [isLoading, setIsLoading] = useState(false);

    const classes = mergeClasses(defaultClasses, props.classes);
    const mask = isLoading ? <div className={classes.mask} /> : null;

    useEffect(() => {
        // After the imageRef mounts, get its --image-width so
        // we can pass it to `resourceUrl`.
        const imageWidth = getComputedStyle(imageRef.current).getPropertyValue(
            '--image-width'
        );

        if (imageWidth) {
            // parseFloat to strip off any units like 'px'.
            const parsedImageWidth = parseFloat(imageWidth);
            setImageSource(
                `${resourceUrl(image.file, {
                    type: 'image-product',
                    width: parsedImageWidth
                })}`
            );
        }
    }, [imageRef]);

    const handleFavoriteItem = useCallback(() => {
        setIsFavorite(!isFavorite);
    }, [isFavorite, setIsFavorite]);
    const handleEditItem = useCallback(() => {
        beginEditItem(item);
    }, [beginEditItem, item]);
    const handleRemoveItem = useCallback(() => {
        setIsLoading(true);

        // TODO: prompt user to confirm this action?
        removeItemFromCart({ item });
    }, [item, removeItemFromCart, setIsLoading]);

    return (
        <li className={classes.root}>
            <img
                alt={name}
                className={classes.image}
                ref={imageRef}
                src={imageSource}
            />
            <div className={classes.name}>{name}</div>
            <ProductOptions options={options} />
            <div className={classes.quantity}>
                <div className={classes.quantityRow}>
                    <span>{qty}</span>
                    <span className={classes.quantityOperator}>{'Ã—'}</span>
                    <span className={classes.price}>
                        <Price currencyCode={currencyCode} value={price} />
                    </span>
                </div>
            </div>
            {mask}
            <Kebab>
                <Section
                    text="Add to favorites"
                    onClick={handleFavoriteItem}
                    icon="Heart"
                    iconAttributes={isFavorite ? FAVORITES_FILL : {}}
                />
                <Section
                    text="Edit item"
                    onClick={handleEditItem}
                    icon="Edit2"
                />
                <Section
                    text="Remove item"
                    onClick={handleRemoveItem}
                    icon="Trash"
                />
            </Kebab>
        </li>
    );
};

Product.propTypes = {
    beginEditItem: func.isRequired,
    currencyCode: string,
    item: shape({
        image: shape({
            file: string.isRequired
        }).isRequired,
        name: string,
        options: array,
        price: number,
        qty: number
    }).isRequired,
    removeItemFromCart: func.isRequired
};

export default Product;
