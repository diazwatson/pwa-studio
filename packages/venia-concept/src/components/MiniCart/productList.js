import React from 'react';
import { arrayOf, func, number, shape, string } from 'prop-types';
import { List } from '@magento/peregrine';

import { mergeClasses } from 'src/classify';
import getCurrencyCode from 'src/util/getCurrencyCode';

import Product from './product';
import defaultClasses from './productList.css';

const ProductList = props => {
    const { beginEditItem, cart, removeItemFromCart } = props;
    const { items } = cart.details;

    const classes = mergeClasses(defaultClasses, props.classes);
    const currency = getCurrencyCode(cart);

    return (
        <List
            classes={classes}
            getItemKey={item => item.item_id}
            items={items}
            render="ul"
            renderItem={props => {
                return (
                    <Product
                        beginEditItem={beginEditItem}
                        currencyCode={currency}
                        item={props.item}
                        removeItemFromCart={removeItemFromCart}
                    />
                );
            }}
        />
    );
};

ProductList.propTypes = {
    beginEditItem: func,
    cart: shape({
        details: shape({
            currency: shape({
                quote_currency_code: string
            }).isRequired,
            items: arrayOf(
                shape({
                    item_id: number,
                    name: string,
                    price: number,
                    product_type: string,
                    qty: number,
                    quote_id: string,
                    sku: string
                })
            )
        }).isRequired
    }).isRequired,
    classes: shape({
        root: string
    }),
    removeItemFromCart: func
};

export default ProductList;
