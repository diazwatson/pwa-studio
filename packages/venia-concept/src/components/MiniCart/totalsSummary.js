import React from 'react';
import { number, object, shape, string } from 'prop-types';

import { Price } from '@magento/peregrine';

import { mergeClasses } from 'src/classify';
import getCurrencyCode from 'src/util/getCurrencyCode';

import defaultClasses from './totalsSummary.css';

const TotalsSummary = props => {
    // Props.
    const { cart } = props;

    // Members.
    const cartCurrencyCode = getCurrencyCode(cart);
    const classes = mergeClasses(defaultClasses, props.classes);
    const hasSubtotal = Boolean(cart.totals.subtotal);
    const itemsQuantity = cart.details.items_qty;
    const itemQuantityText = itemsQuantity === 1 ? 'item' : 'items';
    const totalPrice = cart.totals.subtotal;

    return (
        <div className={classes.root}>
            {hasSubtotal && (
                <dl className={classes.totals}>
                    <dt className={classes.subtotalLabel}>
                        <span>
                            Cart Total :&nbsp;
                            <Price
                                currencyCode={cartCurrencyCode}
                                value={totalPrice}
                            />
                        </span>
                    </dt>
                    <dd className={classes.subtotalValue}>
                        ({itemsQuantity} {itemQuantityText})
                    </dd>
                </dl>
            )}
        </div>
    );
};

TotalsSummary.propTypes = {
    cart: shape({
        details: shape({
            currency: shape({
                quote_currency_code: string
            }).isRequired,
            items_qty: number,
            totals: object
        }).isRequired,
        totals: shape({
            subtotal: number
        }).isRequired
    }).isRequired,
    classes: shape({
        root: string,
        subtotalLabel: string,
        subtotalValue: string,
        totals: string
    })
};

export default TotalsSummary;
