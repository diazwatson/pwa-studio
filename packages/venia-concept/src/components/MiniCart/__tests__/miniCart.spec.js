import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import MiniCart from '../miniCart';

const renderer = new ShallowRenderer();

test('renders the correct tree', () => {
    const props = {
        beginEditItem: jest.fn(),
        endEditItem: jest.fn(),
        cart: {
            isEditingItem: false
        }
    };

    const tree = renderer.render(<MiniCart {...props} />);

    expect(tree).toMatchSnapshot();
});

test('doesnt render a footer when cart is empty', () => {
    const props = {
        beginEditItem: jest.fn(),
        cart: {
            isEditingItem: false
        },
        endEditItem: jest.fn(),
        isCartEmpty: true
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[2];
    expect(footer).toBeNull();
});

test('doesnt render a footer when cart is editing', () => {
    const props = {
        beginEditItem: jest.fn(),
        cart: {
            isEditingItem: true
        },
        endEditItem: jest.fn()
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[2];
    expect(footer).toBeNull();
});

test('doesnt render a footer when cart is loading', () => {
    const props = {
        beginEditItem: jest.fn(),
        cart: {
            isLoading: true
        },
        endEditItem: jest.fn()
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[2];
    expect(footer).toBeNull();
});
