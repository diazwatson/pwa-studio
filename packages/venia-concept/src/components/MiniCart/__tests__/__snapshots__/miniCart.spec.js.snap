// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<aside>
  <Header
    isEditingItem={false}
  />
  <Body
    beginEditItem={[MockFunction]}
    cart={
      Object {
        "isEditingItem": false,
      }
    }
    endEditItem={[MockFunction]}
  />
  <MiniCartMask />
  <Footer
    cart={
      Object {
        "isEditingItem": false,
      }
    }
  />
</aside>
`;
